src/main/java/com/google/sps/servlets/CalendarServlet.java

```diff
@@ -13,28 +23,68 @@
  * Servlet for getting or creating the Calendar page for each group
  */
 @WebServlet("/calendar")
-public class CalendarServlet extends HttpServlet {
-
+public class CalendarServlet extends AbstractEventsServlet {
   /**
-   * Display the Calendar page for each group.
+   * Display the Calendar page for each group. Query string must contain groupId.
    */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    SoyFileSet sfs = SoyFileSet
-        .builder()
-        .add(new File("../../src/main/java/templates/calendar.soy"))
-        .build();
-    SoyTofu tofu = sfs.compileToTofu();
-
-    String out = tofu.newRenderer("templates.calendar.calendarPage").render();
-    response.getWriter().println(out);
+    String groupId = getParameter(request, GROUP_ID_PROPERTY);
+
+    if(!Strings.isNullOrEmpty(groupId)){
+      String out = SoyRendererUtils.getOutputString(CALENDAR_SOY_FILE, CALENDAR_TEMPLATE_NAMESPACE, null);
+
+      response.setContentType("text/html");
+      response.getWriter().println(out);
+    } else {
+      ServletUtils.printBadRequestError(response, CALENDAR_BAD_REQUEST_MESSAGE);
+    }
   }
 
   /**
-   * Create the Calendar page for the group being created
+   * Create a Calendar for the group being created. Query string must contain groupId.
    */
+  @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    //TODO: Implement once groups are implemented
-    throw new UnsupportedOperationException();
+    String groupId = getParameter(request, GROUP_ID_PROPERTY);
+
+    if(!Strings.isNullOrEmpty(groupId)){
+      try {
+        String calendarId = createCalendar(groupId);
+        setGroupCalendarId(groupId, calendarId);
+
+        response.setContentType("text/plain");
+        response.getWriter().println(calendarId);
+      } catch (Exception entityError) {
+        entityError.printStackTrace();
+        response.getWriter().println(CALENDAR_ENTITY_ERROR_MESSAGE);
+      }
+    } else {
+      ServletUtils.printBadRequestError(response, CALENDAR_BAD_REQUEST_MESSAGE);
+    }
+  }
+
+  /**
+   * Return the Calendar ID of the created calendar.
+   */
+  @VisibleForTesting
+  public String createCalendar(String groupId) throws IOException, EntityNotFoundException {
```

Comment:
I haven't used "visibleForTesting" but does the function need to be public or can it be private and the visibleForTesting allow tests to use it?

Reply:
it's pretty much just an extra comment, doesn't do anything else :(